packages/frontend/src/components/Header.jsx

```diff
@@ -5,23 +5,45 @@ import state from '../contexts/state'
 import './header.css'
 
 export default observer(() => {
+  const { unirep } = useContext(state)
   const navigate = useNavigate()
   const [searchInput, setSearchInput] = useState('')
+  useEffect(() => {
+    const loadData = async () => {
+      await Promise.all([
+        unirep.loadAllSignUps(),
+        unirep.loadAttesterDeployments(),
+        unirep.loadAllAttestations(),
+      ])
+    }
+    loadData()
+  }, [])
 
   const search = () => {
     if (!/^(0x)?[a-fA-F0-9]*$/.test(searchInput)) {
       // invalid input, only accept hexadecimal
       // TODO: highlight the field red or something
       return
     }
-    const val = BigInt(`0x${searchInput.replace('0x', '')}`)
-    if (val > BigInt(2) ** BigInt(160)) {
-      // it's an epoch key
-      navigate(`/user/${searchInput})`)
+    // const val = BigInt(`0x${searchInput.replace('0x', '')}`)
+    // if (val > BigInt(2) ** BigInt(160)) {
+    //   // it's an epoch key
+    //   navigate(`/epochKey/${searchInput}`)
+    // } else {
+    //   // otherwise treat it as an attester id
+    //   navigate(`/attester/${searchInput}`)
+    // }
+    const inputAsId = BigInt(searchInput).toString(10)
+    if (unirep.deploymentsById.has(inputAsId)) {
+      navigate(`/attester/${searchInput}`)
+    } else if (unirep.signUpsByUserId.has(inputAsId)) {
+      navigate(`/user/${searchInput}`)
+    } else if (unirep.attestationsByEpochKey.has(inputAsId)) {
+      navigate(`/epochKey/${searchInput}`)
     } else {
-      // otherwise treat it as an attester id
-      navigate(`/attester/${searchInput})`)
+      navigate(`/NotFound`)
     }
+    setSearchInput('')
   }
```

Comment:
so the original check `if (val > BigInt(2) ** BigInt(160))` was allowing any value too small to be an epoch key or user to render `AttesterPage` with that arbitrary value as an attesterId.. how do you feel about this way instead?
I added another mapping to the ingestion of allAttestations in `unirep.js`, so now we could use `unirep.attestationsByEpochKey` to render data for EpochKeyPage and do away with the epochKey context.

Reply:
This isn't something we can/should solve on the client side. Generally you shouldn't assume that you have access to _all_ attestations/signups/attesters/etc in the client data. This is an anti-pattern that is ok to bootstrap the ui, but should be removed as we go.

Also, note that an epoch key/user id _could_ be `< 2**160`, the odds are just quite small. It would be like finding an ethereum address with 10 zeroes in it.

Reply:
> To tell the difference we need to ask the server whether the input is an `epochKey` or `userId` (or neither). The server will check if it has seen the identifier in either a signup event, or an attestation event.

ok seems I misunderstood this. Do you mean that `search()` should call a new function, maybe `unirep.findId()`, that sends a request to check both`Attestation` and `UserSignUp` tables for the `searchInput`? .. and that function could return the type of the id it finds?

Reply:
Sorta, you should think of the frontend and backend as separate entities. The frontend has an idea of what an attestation or signup is, but doesn't know/care about the database structure, that's the responsibility of the backend. So the frontend will call a route like `/api/search` and provide the search string. The backend might then return a structure like

```js
{
  type: 'attester' | 'epochKey' | 'userId' | 'unknown',
  data: {} // the full object (if needed)
}
```

Based on this response you can ingest the `data` portion of the response into the relevant store and then navigate to the destination page (where the data will be populated).

Reply:
But yes, internally the `/api/search` function would search the `Attestation`/`UserSignUp` tables. Sorry just wanted to clarify the frontend/backend logical separation.

Reply:
ok yes I get it thank you. 
So I only returned the `type` in the response.  I don't think I need to use any data from the object used to find it; relevant data gets loaded when `search() `navigates to the new page.

Reply:
Right, but if you only include `type` in the search response the user has to sit through two loading times. First the search request to get the `type` info, then the actual data loading when the user gets to the relevant page. If you include the data object in the search response you can pre-populate the store so that the user only experiences one loading time.